nemo/collections/vlm/neva/data/lazy.py

```diff
@@ -0,0 +1,612 @@
+    template_len = len(template)
+    pattern_len = len(pattern)
+    for i in range(search_start_index, template_len - pattern_len + 1):
+        match = template[i : i + pattern_len] == pattern
+        if torch.all(match) or (allow_first_token_mismatch and torch.all(match[1:])):
+            return i, i + pattern_len
+    return -1, -1
+
+
+class LazySupervisedDataset(Dataset):
+
+    def __init__(
+        self,
+        data_path,
+        data_config,
+        tokenizer,
+        image_processor,
+    ):
+        super().__init__()
+        if data_path is not None:
+            with open(data_path, "r") as file:
+                list_data_dict = json.load(file)
+        else:
+            list_data_dict = []
+
+        logging.warning("Formatting inputs...Skip in lazy mode")
+        self.data_config = data_config
+        self.tokenizer = tokenizer
+        self.image_processor = image_processor
+
+        self.conv_template = data_config.conv_template
+        self.conv = supported_conv_templates[self.conv_template]
+        self.image_process_mode = data_config.image_process_mode
+        self.list_data_dict = list_data_dict
+
+        image_folder = getattr(data_config, "image_folder", None)
+        video_folder = getattr(data_config, "video_folder", None)
+
+        self.image_loader = TarOrFolderImageLoader(image_folder) if image_folder else None
+        self.video_loader = TarOrFolderVideoLoader(video_folder, data_config) if video_folder else None
+
+    def __len__(self):
+        return len(self.list_data_dict)
+
+    def __getitem__(self, i) -> Dict[str, torch.Tensor]:
+        source = self.list_data_dict[i]
+        conversations = self._apply_prompt_templates(source, use_plain=self.conv_template == "plain")
+        tokens, labels = self._tokenize_and_label(conversations)
+
+        media_tensors = self._process_images(source)
+        data_dict = dict(
+            image=media_tensors,
+            tokens=tokens,
+            labels=labels,
+        )
+        return data_dict
+
+    def _process_images(self, source):
+        media_tensors = torch.tensor([])
+        if 'image' in source:
+            if not isinstance(source['image'], list):
+                source['image'] = [source['image']]
+
+            images = []
+            for image_file in source['image']:
+                image = self.image_loader.open_image(image_file)
+                if image is None:
+                    logging.warning(f"Image {image_file} could not be found!")
+                image = process_image(self.image_processor, image, self.image_process_mode)
+                images.append(image)
+
+            if images:
+                media_tensors = torch.stack(images)
+        return media_tensors
+
+    def _apply_prompt_templates(self, source, use_plain=False):
+        conv = self.conv
+
+        roles = {"human": conv.roles[0], "gpt": conv.roles[1]}
```

Comment:
For pretraining data using plain template, do we still support the jsonl format we used before, where it is "User" "Assistant" instead of human/gpt?

Reply:
I think I manually converted the user/assitant to human/gpt before for jsonl files